### ISSUE TITLE
Window drag redraw has high latency

### ISSUE THREAD

**Issue by TurkeyMan** [linuxmint/Cinnamon#4969]
Switching between Windows and Mint, it's very obvious that dragging is not smooth and snappy.
In Windows, the window stays firmly fixed to the mouse cursor, ie, move the mouse cursor, the window moves, and the tracks the exact pixel the mouse clicked on.
In Mint, the windows move visibly 1-2 frames behind the mouse cursor. The cursor moves off ahead of the window, and stop moving the mouse, you can watch a couple of frames where the window 'catches up' with the mouse. I estimate 2, maybe 3 frames of rendering latency.

My computer is very fast, it is not a resource problem, and this happens the same on every computer I've ever tried.

This may be related to: https://github.com/linuxmint/Cinnamon/issues/2465
But I think this is a separate issue, since I also experience the stutter mentioned in that issue as an additional problem.

**Comment by ManIVIctorious** [comment]
Unfortunately this is an issue of the underlying displayserver X11, so the problem won't be resolved till cinnamon switches to newer technologies like wayland

**Comment by TurkeyMan** [comment]
Really? Is that on the roadmap? Wayland has been in the wings for years and years... is it actually coming? :)

How does X11 imply a couple of frames latency?

**Comment by ManIVIctorious** [comment]
As far as I know there haven't been any plans yet, but someday it will have to come (although I'm pretty sure initial support won't happen within the next year).

By not keeping anything in sync. Neither a mousepointer with a window, nor a pluginwindow inside of a browser. X11 is all about tearing

**Comment by lestcape** [comment]
I really disagree with @ManIVIctorious about x11. Or other desktop environment, different of gnome shell and cinnamon run in what windows manager? The problem in my opinion is how the windows are embedded inside cinnamon, to be possible controlled it from javascript. There are a delay always on all thing that you execute. In this case, the delay is big so you can see the effect. Is not the only one but thing about how the windows are embedded and in my opinion this is not the worse. Try to open and then close several videos and you will see a big performance degradation.

**Comment by TurkeyMan** [comment]
@lestcape Right, what you say is exactly what I imagined by how it looks to me. I think architecture leading to frames of delay (from event driven or scripted systems, or whatever) should be considered a bug in something like a window manager/desktop environment.
How many realtime Cinnamonm dev's are on the team? Any game developers? Is there tooling available to give profiling metrics like issuing redraw times, rendering times, presentation time, etc?
I suspect X11 plays a part, and Wayland might reduce maybe one of those frames of latency, but I can easily see ~3 frames of latency, and I find it hard to believe X is responsible for all of them.

**Comment by lestcape** [comment]
@TurkeyMan That I know, there are not any a full time dev in cinnamon and also not a game developer.

This problem is relate with the compatibility, relate with the windows manager, and relate with Cutter, and more things. Probably this could be better with Wayland, because the client-server architecture of Wayland probably could give a way to allow the interception of a different application/library like Clutter and then decide a group of things directly without need all current layers to communicate the js part with the xwindows system.

This problem is also known by the clutter devs, because they have an special way to handle some situations like that, when you need more performance. This is the case of https://people.gnome.org/~gcampagna/docs/Clutter-1.0/Clutter.grab_pointer.html. What this function can do is pass the control of the mouse to the actor, because they know is the only way to received the events. i.e. they know about the mouse can be in other different position of the actor in a drag and drop procedure. So, I doubt the cinnamon developers can do something about it that escapes beyond its scope. They do not develop the clutter API.

**Comment by ManIVIctorious** [comment]
@lestcape the not synced mouse pointer is a long standing issue of X.org-server. 
You can compare gnome on X11 and gnome on wayland to see the difference there. I don't really know, but I think that gnome and cinnamon handle window drag quite similar.

**Comment by TurkeyMan** [comment]
There's no additional JS event in the chain or anything? If I had any idea how the code looked, I would check that there are zero points where any input message is *queued* in between mouse input and window movement response. Input queues can often be responsible for a frame of latency each time an event or response hits a dispatch queue. Message pump's are okay for clicks, but not for mouse move/drag's. I have a feeling there's a lot of queuing happening around the place. There are places all over cinnamon where response has much longer latency than I would expect, but dragging windows is particularly annoying and immediately obvious to anyone coming from any other OS.

**Comment by lestcape** [comment]
@ManIVIctorious, i think we are agree but we have a different perspective of the cause. If you design a system to work in one way, and then other people think that is ok do a different thing, they do it and see that not work as they want then, what it's the cause? The original implementation or the new invention?

**Comment by lestcape** [comment]
@TurkeyMan of course there are a lot of js events involved on this. You need to inform that the situation changed. The move not occurs outside the stage (the cinnamon stage), so, the stage receive the mouse move, is also who send the message to move the windows and then received again the new windows position.

All begin here:
https://github.com/linuxmint/Cinnamon/blob/master/js/ui/windowManager.js#L405

but you will found what you're looking for here:
https://github.com/linuxmint/Cinnamon/blob/master/js/ui/xdndHandler.js

**Comment by TurkeyMan** [comment]
I see. How much queuing of events is going on?
I've never written js before, but browsing through, it looks like the event handlers are called immediately when signals emit, so that shouldn't introduce latency. Latency would only normally be introduced when messages are deferred or queued and handled in a traditional message pump style. So maybe that's not happening here?

**Comment by lestcape** [comment]
I don't know how much queuing of events is going on, i never tracked all this... Any way what i can tell you is that javascript only have one thread, so, the event are queuing of course and that also could have priority for some events, but this is not implemented in js, maybe in clutter. Look, i have a phd in combinatorial optimization. Thats is not useful here when we speak of a real program in a real language with the intervention of a lot of libraries and more things. Here what is difficult (at less to me) is know what you are asking. If i know where is the problem be sure i know how make an algorithm to reduce the computational order.

**Comment by TurkeyMan** [comment]
I'm not really asking anything, only the report in the initial bug.

I was only asking about event buffering/queueing, because after 15 years of experience in realtime systems, this is by far the most common cause for a problem like this that I see over and over again. It was just an educated guess, and I thought somebody that knows the cinnamon windowing code might be able to make comment on the structure, and if any such event queues exist that might contribute to the problem. Don't worry about my questions though, they're obviously not helpful.

**Comment by lestcape** [comment]
@TurkeyMan i think you understand wrong, like you try to find the origin of this, not? So, i can not help you on this, because i don't know. I also not speak english as you can see, sorry. Where i say: "Here what is difficult (at less to me) is know what you are asking.", what i try to say is "Here what is difficult (at less to me) is know where is the problem (localize the problem)." Again sorry.

**Comment by TurkeyMan** [comment]
It's all good, thanks for taking the time to answer me.

**Comment by dandv** [comment]
Same issue with Cinnamon on Mint 18 Sarah.

**Comment by icarter09** [comment]
@TurkeyMan are you still experiencing this issue?
